File: ms-pedidos/src/main/java/com/smartlogix/ms_pedidos/controller/PedidoController.java
package com.smartlogix.ms_pedidos.controller;

import com.smartlogix.ms_pedidos.dto.PedidoDTO;
import com.smartlogix.ms_pedidos.model.EstadoPedido;
import com.smartlogix.ms_pedidos.model.Pedido;
import com.smartlogix.ms_pedidos.service.PedidoService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/pedidos")
@RequiredArgsConstructor
public class PedidoController {

    private final PedidoService pedidoService;

    @GetMapping
    public ResponseEntity<List<Pedido>> listarTodos() {
        return ResponseEntity.ok(pedidoService.listarTodos());
    }

    @GetMapping("/{id}")
    public ResponseEntity<Pedido> obtenerPorId(@PathVariable Long id) {
        return ResponseEntity.ok(pedidoService.obtenerPorId(id));
    }

    @PostMapping
    public ResponseEntity<Pedido> crearPedido(@RequestBody PedidoDTO dto) {
        Pedido pedidoCreado = pedidoService.crearPedido(dto);
        return ResponseEntity.status(HttpStatus.CREATED).body(pedidoCreado);
    }

    @PutMapping("/{id}/estado")
    public ResponseEntity<Pedido> actualizarEstado(@PathVariable Long id, @RequestBody Map<String, String> body) {
        String estadoStr = body.get("estado");
        if (estadoStr == null) {
            throw new IllegalArgumentException("El campo 'estado' es requerido");
        }
        EstadoPedido nuevoEstado = EstadoPedido.valueOf(estadoStr.toUpperCase());
        return ResponseEntity.ok(pedidoService.actualizarEstado(id, nuevoEstado));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Pedido> cancelarPedido(@PathVariable Long id) {
        return ResponseEntity.ok(pedidoService.actualizarEstado(id, EstadoPedido.CANCELADO));
    }

    @GetMapping("/cliente/{clienteId}")
    public ResponseEntity<List<Pedido>> listarPorCliente(@PathVariable Long clienteId) {
        return ResponseEntity.ok(pedidoService.listarPorCliente(clienteId));
    }

    @GetMapping("/estado/{estado}")
    public ResponseEntity<List<Pedido>> listarPorEstado(@PathVariable EstadoPedido estado) {
        return ResponseEntity.ok(pedidoService.listarPorEstado(estado));
    }
}

File: ms-pedidos/src/pedidos/pedido.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  HttpCode,
  HttpStatus,
  ParseIntPipe,
} from '@nestjs/common';
import { PedidoService } from './pedido.service';
import { PedidoDTO } from './pedido.dto';
import { Pedido, EstadoPedido } from './pedido.entity';
import { IllegalArgumentException } from './exceptions';

@Controller('pedidos')
export class PedidoController {
  constructor(private readonly service: PedidoService) {}

  @Get()
  async listarTodos(): Promise<Pedido[]> {
    return this.service.listarTodos();
  }

  @Get(':id')
  async obtenerPorId(@Param('id', ParseIntPipe) id: number): Promise<Pedido> {
    return this.service.obtenerPorId(id);
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  async crearPedido(@Body() dto: PedidoDTO): Promise<Pedido> {
    return this.service.crearPedido(dto);
  }

  @Put(':id/estado')
  async actualizarEstado(
    @Param('id', ParseIntPipe) id: number,
    @Body() body: Record<string, string>,
  ): Promise<Pedido> {
    const estadoStr = body.estado;
    if (!estadoStr) {
      throw new IllegalArgumentException("El campo 'estado' es requerido");
    }

    const upperEstado = estadoStr.toUpperCase();
    if (!Object.values(EstadoPedido).includes(upperEstado as EstadoPedido)) {
      throw new IllegalArgumentException(`Estado inválido: ${estadoStr}`);
    }

    return this.service.actualizarEstado(id, upperEstado as EstadoPedido);
  }

  @Delete(':id')
  async cancelarPedido(@Param('id', ParseIntPipe) id: number): Promise<Pedido> {
    return this.service.actualizarEstado(id, EstadoPedido.CANCELADO);
  }

  @Get('cliente/:clienteId')
  async listarPorCliente(@Param('clienteId', ParseIntPipe) clienteId: number): Promise<Pedido[]> {
    return this.service.listarPorCliente(clienteId);
  }

  @Get('estado/:estado')
  async listarPorEstado(@Param('estado') estado: string): Promise<Pedido[]> {
    const upperEstado = estado.toUpperCase();
    if (!Object.values(EstadoPedido).includes(upperEstado as EstadoPedido)) {
      throw new IllegalArgumentException(`Estado inválido: ${estado}`);
    }
    return this.service.listarPorEstado(upperEstado as EstadoPedido);
  }
}
